R/DataInputPanel.R

```diff
@@ -0,0 +1,197 @@
+                     incon = icon("trash"),
+                     style = "color: #fff; background-color: #dc3545; border-color: #dc3545")
+      )
+    ),
+    div(class = "clearfix")
+  )
+}
+
+#' Find all of the treatment names in the data, both for long and wide formats.
+#' 
+#' @param data Data frame in which to search for treatment names
+#' @return Vector of all treatment names
+find_all_treatments <- function(data) {
+  if ('T' %in% colnames(data)) {
+    # Long format
+    return(unique(data$T))
+  } else {
+    # Wide format
+    all <- c()
+    for (col in paste0('T.', seq(6))) {
+      if (col %in% colnames(data)) {
+        all <- c(all, data[[col]])
+      }
+    }
+    return(unique(all[!is.na(all)]))
+  }
+}
+
+#' Create a copy of a vector with the given item as the first element.
+#' 
+#' @param vector Vector to reorder
+#' @param first_item The element to push to the front of the vector
+#' @return The reordered vector
+vector_with_item_first <- function(vector, first_item) {
+  if (is.null(first_item) || !(first_item %in% vector)) {
+    return(vector)
+  }
+  return(c(first_item, vector[vector != first_item]))
+}
+
+# Treatments are in priority order, such that or any study with multiple matching treatments,
+# the first in this vector will be used as the reference.
+potential_reference_treatments = c(
+  'control',
+  'usual_care',
+  'standard_care',
+  'placebo',
+  'no_contact'
+)
+
+#' Find the expected reference treatment from a vector.
+#' This is done by comparing treatment names to expected reference treatment names.
+#' 
+#' @param treatments vector containing all treatment names
+#' @return Name of the expected reference treatment if one is found, else NULL
+find_expected_reference_treatment <- function(treatments) {
+  expected_reference_treatments <- unlist(
+    lapply(potential_reference_treatments,
+           function(expected_ref) {
+             match(expected_ref, tolower(treatments))
+           }))
+  expected_reference_treatments <- expected_reference_treatments[!is.na(expected_reference_treatments)]
+  if (length(expected_reference_treatments) > 0) {
+    return(treatments[expected_reference_treatments[1]])
+  } else {
+    return(NULL)
+  }
+}
+
+#' Module server for uploading data into the app.
+#' 
+#' @param id ID of the module
+#' @param metaoutcome Reactive containing the outcome type selected
+#' @return List of reactives:
+#'   - 'data' is the uplodaded data
+#'   - 'treatment_list' is the data frame containing the treatment ID ('Number') and the treatment name ('Label')
+data_input_panel_server <- function(id, metaoutcome) {
+  moduleServer(id, function(input, output, session) {
+    ns <- session$ns
+    
+    # Create a definable reactive value to allow reloading of data
+    reload <- reactiveVal(FALSE)
+
+    # Render function for file input dynamically to allow the button to be set to Null
+    default_file_input <- renderUI({
+      fileInput(
+        inputId = ns("data"),
```

Comment:
Yep, this is only needed because this is in a `renderUI()` block. In the *Dynamic UI* section that you linked, they do use the `NS()` function, though it's not entirely trivial to follow.